[opencv.js threads] build opencv.js with pthreads

Wenzhao-Xiang opened the pull request:
<!-- Please use this line to close one or multiple issues when this pullrequest gets merged
You can add another line right under the first one:
resolves #1234
resolves #1235
-->

### This pullrequest changes
First try of building opencv.js with pthreads. 

### Details
Please build with the command:
 ```sh
python ./platforms/js/build_js.py build_wasm --build_wasm --build_test
```
After building, you will get three files which are `opencv.js`, `opencv_js.js` and `opencv_js.worker.js`. The first is the cv module with wrapper and the second is the cv module without wrapper. And the last is the web worker definition file. 

Please change Line 109 in `opencv_js.worker.js` from `importScripts('e.data.urlOrBlob')` to `importScripts('./opencv_js.js')`, because web worker can't import opencv.js correctly with wrapper. After that change, web worker can be allocated correctly.

### Issues
Now the pthreads version of opencv.js still doesn't work. In my investigation, when I test the `resize` kernel, all the calculations were done by main thread, and no no web worker was used. Seems we need investigate this issue further.
<!-- Please describe what your pullrequest is changing -->

Wenzhao-Xiang committed:
[opencv.js threads] build opencv.js with pthreads

Wenzhao-Xiang commented:
@huningxin @terfendail Please help take a look. Thanks!

Wenzhao-Xiang commented:
@huningxin @terfendail fix the pthreads issue.
It is because opencv set the default threads number to 1. I changed the code to set the thread number to 4 forcely, and then found all things works normally. Seems there are about 2-3x speed up in benchmark test. I will next:
1. collect detailed data for opencv.js threads
2. investigate how to set thread number according to CPU cores number.

Wenzhao-Xiang commented:
I updated the [performance sheet](https://docs.google.com/spreadsheets/d/17TS4wBDuKUr2aN_dOPlsGDX_7Zh8_XlUU3ownv3VZHs/edit#gid=0).

huningxin commented:
> Seems there are about 2-3x speed up in benchmark test.

Fantastic! Thanks @Wenzhao-Xiang .

> It is because opencv set the default threads number to 1.

Do you know the root cause of this behavior?

> I changed the code to set the thread number to 4 forcely, and then found all things works normally.

What's your change? I don't see that in your PR.

Wenzhao-Xiang committed:
[opencv.js threads] set threads number to 4 forcely

Wenzhao-Xiang commented:
> Do you know the root cause of this behavior?

@huningxin 
I see in https://github.com/huningxin/opencv/blob/c396120d651121afac2853dc16402a636615b67b/modules/core/src/parallel.cpp#L667, opencv set the default threads number to [`default_number_of_threads`](https://github.com/huningxin/opencv/blob/c396120d651121afac2853dc16402a636615b67b/modules/core/src/parallel.cpp#L674) or [`config_num_threads`](https://github.com/huningxin/opencv/blob/c396120d651121afac2853dc16402a636615b67b/modules/core/src/parallel.cpp#L679), I'm now investigating which is the root cause.



> What's your change? I don't see that in your PR.

Just push the lastest commit.

Wenzhao-Xiang commented:
> I see in
> 
> https://github.com/huningxin/opencv/blob/c396120d651121afac2853dc16402a636615b67b/modules/core/src/parallel.cpp#L667
> 
> , opencv set the default threads number to [`default_number_of_threads`](https://github.com/huningxin/opencv/blob/c396120d651121afac2853dc16402a636615b67b/modules/core/src/parallel.cpp#L674) or [`config_num_threads`](https://github.com/huningxin/opencv/blob/c396120d651121afac2853dc16402a636615b67b/modules/core/src/parallel.cpp#L679), I'm now investigating which is the root cause.

Seems it's because `cv::getNumberOfCPUs()` doesn't work in WASM and just return the default value 1. We may need to expose the method of setting threads number.

Wenzhao-Xiang committed:
[opencv.js threads] expose threads number setting method

Wenzhao-Xiang commented:
I test some cases with `pthread` + `allow_memory_growth`, and found there seems no performance issue with these flags.
https://github.com/opencv/opencv/issues/14691#issuecomment-509899734
https://github.com/opencv/opencv/issues/14691#issuecomment-509904489

huningxin commented:
> Seems it's because `cv::getNumberOfCPUs()` doesn't work in WASM and just return the default value 1

What's the underlying browser API it calls? There is [`navigator.hardwareConcurrency`](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorConcurrentHardware/hardwareConcurrency) that provides the number of logical cores.

huningxin commented:
> I test some cases with `pthread` + `allow_memory_growth`, and found there seems no performance issue with these flags.

I would suggest that probably we need to introduce a build flag, e.g. `--threads`, to guard the multi-threads enabled build. 

@terfendail , WDYT?

Wenzhao-Xiang commented:
> What's the underlying browser API it calls? There is [`navigator.hardwareConcurrency`](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorConcurrentHardware/hardwareConcurrency) that provides the number of logical cores.

See here: https://github.com/huningxin/opencv/blob/c396120d651121afac2853dc16402a636615b67b/modules/core/src/parallel.cpp#L817

 Seems there is no browser environment being detected and it just go to `else` option which returns `1`.

huningxin commented:
I see. 

As https://github.com/emscripten-core/emscripten/blob/incoming/src/library.js#L464, emscripten actually supports the `sysconf( _SC_NPROCESSORS_ONLN )`. Let's try to add `__EMSCRIPTEN__` into `#elif defined __linux__ || defined __GLIBC__ || defined __HAIKU__` and try again.

Wenzhao-Xiang commented:
> As https://github.com/emscripten-core/emscripten/blob/incoming/src/library.js#L464, emscripten actually supports the `sysconf( _SC_NPROCESSORS_ONLN )`. Let's try to add `__EMSCRIPTEN__` into `#elif defined __linux__ || defined __GLIBC__ || defined __HAIKU__` and try again.

Sounds good! Thanks! I'll try it.

Wenzhao-Xiang commented:
@huningxin 
It works! Btw, Do we still need to expose the method of setting threads number?

Wenzhao-Xiang committed:
[opencv.js threads] 

1. adapt cv::getNumberOfCPUs to browser
2. adapt opencv.js wrapper to web worker

huningxin commented:
> It works!

I am happy to hear. Nice job.

> Do we still need to expose the method of setting threads number?

I suppose this is a good feature that allows developers to control the concurrency.

huningxin commented:
> I would suggest that probably we need to introduce a build flag, e.g. `--threads`, to guard the multi-threads enabled build.

Any update on this?

> > Do we still need to expose the method of setting threads number?
> 
> I suppose this is a good feature that allows developers to control the concurrency.

and this?

terfendail commented on modules/js/CMakeLists.txt:
```diff
@@ -91,8 +91,8 @@ endif()
-set(EMSCRIPTEN_LINK_FLAGS "${EMSCRIPTEN_LINK_FLAGS} --memory-init-file 0 -s TOTAL_MEMORY=134217728 -s ALLOW_MEMORY_GROWTH=1")
-set(EMSCRIPTEN_LINK_FLAGS "${EMSCRIPTEN_LINK_FLAGS} -s MODULARIZE=1 -s SINGLE_FILE=1")
+set(EMSCRIPTEN_LINK_FLAGS "${EMSCRIPTEN_LINK_FLAGS} --memory-init-file 0 -s TOTAL_MEMORY=134217728 -s ALLOW_MEMORY_GROWTH=0")
+set(EMSCRIPTEN_LINK_FLAGS "${EMSCRIPTEN_LINK_FLAGS} -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=4 -s MODULARIZE=1 -s SINGLE_FILE=1")
```
I suppose selection of thread related flags should depend on `HAVE_PTHREADS_PF` cmake flag that depends on 'WITH_PTHREADS_PF` configuration value. So it would be possible to configure `WITH_PTHREADS_PF` in accordance with `--threads` flag for `build_js.py`.

Wenzhao-Xiang replied:
Thanks! Implemented here: 5369aea

Wenzhao-Xiang committed:
[opencv.js threads] support `--threads` buildind option

Wenzhao-Xiang committed:
[opencv.js threads] support `USE_PTHREADS`+`ALLOW_MEMORY_GROWTH`

huningxin reviewed (approved):
Looks good. Thanks!

huningxin commented:
I can reproduce the performance speed on my machine. For example, for `CvtColor` color with parameter `(640x480,COLOR_RGBA2GRAY)`, the threading version (4 threads) is **2.7X** faster than scalar version. Nice job @Wenzhao-Xiang !

Only issue of the perf test is https://github.com/huningxin/opencv/issues/269, please help fix that.

terfendail reviewed (approved):
Looks fine for me

huningxin commented:
Thanks for the PR and review. 

@Wenzhao-Xiang , please followup #269 and #266 , so we can run perf test smoothly.
